### ISSUE TITLE
Failed to compile in Windows 10?

### ISSUE THREAD

**Issue by veniamin-ilmer** [john01dav/softbuffer#24]
Hello, this is my first time trying to use your library.

I just copied and pasted your example from the readme:

```
use softbuffer::GraphicsContext;
use winit::event::{Event, WindowEvent};
use winit::event_loop::{ControlFlow, EventLoop};
use winit::window::WindowBuilder;

fn main() {
    let event_loop = EventLoop::new();
    let window = WindowBuilder::new().build(&event_loop).unwrap();
    let mut graphics_context = unsafe { GraphicsContext::new(window) }.unwrap();

    event_loop.run(move |event, _, control_flow| {
        *control_flow = ControlFlow::Wait;

        match event {
            Event::RedrawRequested(window_id) if window_id == graphics_context.window().id() => {
                let (width, height) = {
                    let size = graphics_context.window().inner_size();
                    (size.width, size.height)
                };
                let buffer = (0..((width * height) as usize))
                    .map(|index| {
                        let y = index / (width as usize);
                        let x = index % (width as usize);
                        let red = x % 255;
                        let green = y % 255;
                        let blue = (x * y) % 255;

                        let color = blue | (green << 8) | (red << 16);

                        color as u32
                    })
                    .collect::<Vec<_>>();

                graphics_context.set_buffer(&buffer, width as u16, height as u16);
            }
            Event::WindowEvent {
                event: WindowEvent::CloseRequested,
                window_id,
            } if window_id == graphics_context.window().id() => {
                *control_flow = ControlFlow::Exit;
            }
            _ => {}
        }
    });
}
```

Here is my cargo.toml:
```
[dependencies]
softbuffer = "0.1.1"
winit = "0.27.1"
```

I get this when trying to build:
```
error[E0432]: unresolved import `winapi::shared::windef`
 --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\win32.rs:4:21
  |
4 | use winapi::shared::windef::{HDC, HWND};
  |                     ^^^^^^ could not find `windef` in `shared`

error[E0432]: unresolved import `winapi::um::wingdi`
 --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\win32.rs:5:17
  |
5 | use winapi::um::wingdi::{StretchDIBits, BITMAPINFOHEADER, BI_BITFIELDS, RGBQUAD};
  |                 ^^^^^^ could not find `wingdi` in `um`

error[E0432]: unresolved import `winapi::um::winuser`
 --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\win32.rs:6:17
  |
6 | use winapi::um::winuser::{GetDC, ValidateRect};
  |                 ^^^^^^^ could not find `winuser` in `um`

error[E0433]: failed to resolve: could not find `wingdi` in `um`
  --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\win32.rs:64:25
   |
64 |             winapi::um::wingdi::DIB_RGB_COLORS,
   |                         ^^^^^^ could not find `wingdi` in `um`

error[E0433]: failed to resolve: could not find `wingdi` in `um`
  --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\win32.rs:65:25
   |
65 |             winapi::um::wingdi::SRCCOPY,
   |                         ^^^^^^ could not find `wingdi` in `um`

```

Did I do something wrong?

Otherwise, I see a lot of potential in your library, how we use winit instead of trying to reinvent the wheel.

Thank you,
Veniamin

**Comment by veniamin-ilmer** [comment]
I was able to get around this problem by adding this dependency:
`winapi = { version = "*", features = ["windef", "wingdi", "winuser"] }`

However, the example still won't compile:

```
error[E0277]: the trait bound `Window: raw_window_handle::HasRawWindowHandle` is not satisfied
    --> src\main.rs:9:60
     |
9    |   let mut graphics_context = unsafe { GraphicsContext::new(window) }.unwrap();
     |                                       -------------------- ^^^^^^ the trait `raw_window_handle::HasRawWindowHandle` is not implemented for `Window`
     |                                       |
     |                                       required by a bound introduced by this call
     |
help: trait impl with same name found
    --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.27.1\src\window.rs:1039:1
     |
1039 | / unsafe impl HasRawWindowHandle for Window {
1040 | |     /// Returns a [`raw_window_handle::RawWindowHandle`] for the Window
1041 | |     ///
1042 | |     /// ## Platform-specific
...    |
1057 | |     }
1058 | | }
     | |_^
     = note: perhaps two different versions of crate `raw_window_handle` are being used?
     = help: the trait `raw_window_handle::HasRawWindowHandle` is implemented for `&'a T`
note: required by a bound in `GraphicsContext::<W>::new`
    --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\lib.rs:33:9
     |
33   | impl<W: HasRawWindowHandle> GraphicsContext<W> {
     |         ^^^^^^^^^^^^^^^^^^ required by this bound in `GraphicsContext::<W>::new`

error[E0277]: the trait bound `Window: raw_window_handle::HasRawWindowHandle` is not satisfied
    --> src\main.rs:9:39
     |
9    |   let mut graphics_context = unsafe { GraphicsContext::new(window) }.unwrap();
     |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `raw_window_handle::HasRawWindowHandle` is not implemented for `Window`
     |
help: trait impl with same name found
    --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.27.1\src\window.rs:1039:1
     |
1039 | / unsafe impl HasRawWindowHandle for Window {
1040 | |     /// Returns a [`raw_window_handle::RawWindowHandle`] for the Window
1041 | |     ///
1042 | |     /// ## Platform-specific
...    |
1057 | |     }
1058 | | }
     | |_^
     = note: perhaps two different versions of crate `raw_window_handle` are being used?
     = help: the trait `raw_window_handle::HasRawWindowHandle` is implemented for `&'a T`
note: required by a bound in `GraphicsContext`
    --> C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\softbuffer-0.1.1\src\lib.rs:28:31
     |
28   | pub struct GraphicsContext<W: HasRawWindowHandle> {
     |                               ^^^^^^^^^^^^^^^^^^ required by this bound in `GraphicsContext`

error[E0599]: the method `window` exists for struct `GraphicsContext<Window>`, but its trait bounds were not satisfied
  --> src\main.rs:15:74
   |
15 |       Event::RedrawRequested(window_id) if window_id == graphics_context.window().id() => {
   |                                                                          ^^^^^^ private field, not a method
   |
  ::: C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.27.1\src\window.rs:44:1
   |
44 | pub struct Window {
   | ----------------- doesn't satisfy `Window: raw_window_handle::HasRawWindowHandle`
   |
   = note: the following trait bounds were not satisfied:
           `Window: raw_window_handle::HasRawWindowHandle`

error[E0599]: the method `window` exists for struct `GraphicsContext<Window>`, but its trait bounds were not satisfied
  --> src\main.rs:17:39
   |
17 |           let size = graphics_context.window().inner_size();
   |                                       ^^^^^^ private field, not a method
   |
  ::: C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.27.1\src\window.rs:44:1
   |
44 | pub struct Window {
   | ----------------- doesn't satisfy `Window: raw_window_handle::HasRawWindowHandle`
   |
   = note: the following trait bounds were not satisfied:
           `Window: raw_window_handle::HasRawWindowHandle`

error[E0599]: the method `set_buffer` exists for struct `GraphicsContext<Window>`, but its trait bounds were not satisfied
  --> src\main.rs:34:26
   |
34 |         graphics_context.set_buffer(&buffer, width as u16, height as u16);
   |                          ^^^^^^^^^^ method cannot be called on `GraphicsContext<Window>` due to unsatisfied trait bounds
   |
  ::: C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.27.1\src\window.rs:44:1
   |
44 | pub struct Window {
   | ----------------- doesn't satisfy `Window: raw_window_handle::HasRawWindowHandle`
   |
   = note: the following trait bounds were not satisfied:
           `Window: raw_window_handle::HasRawWindowHandle`

error[E0599]: the method `window` exists for struct `GraphicsContext<Window>`, but its trait bounds were not satisfied
  --> src\main.rs:39:42
   |
39 |       } if window_id == graphics_context.window().id() => {
   |                                          ^^^^^^ private field, not a method
   |
  ::: C:\Users\ilmer\.cargo\registry\src\github.com-1ecc6299db9ec823\winit-0.27.1\src\window.rs:44:1
   |
44 | pub struct Window {
   | ----------------- doesn't satisfy `Window: raw_window_handle::HasRawWindowHandle`
   |
   = note: the following trait bounds were not satisfied:
           `Window: raw_window_handle::HasRawWindowHandle`

```

**Comment by n5y** [comment]
Downgrading winit to version 0.26.1 fixes HasRawWindowHandle trait issue.

From winit 0.27.0 (2022-07-26) changelog:
**Breaking**, update `raw-window-handle` to `v0.5` and implement `HasRawDisplayHandle` for `Window` and `EventLoopWindowTarget`.

From raw-window-handle 0.5.0 (2022-07-14) changelog:
**Breaking:** The `RawWindowHandle` variants were split into `RawDisplayHandle` and `RawWindowHandle`.

**Comment by john01dav** [comment]
Fixed in ea6f97e15e012684eac0ba6a65b75f065d2c18f8, thanks to Liamolucko.
